src/librustc_mir/interpret/operator.rs

```diff
@@ -314,13 +314,23 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> InterpretCx<'a, 'mir, 'tcx, M>
                     right.to_scalar_ptr()?.is_ptr() ||
                     bin_op == mir::BinOp::Offset
                 {
-                    return M::ptr_op(self, bin_op, left, right);
+                    if left.layout.ty.is_integral() && right.layout.ty.is_integral() {
```

Comment:
why was this change done? It seems good, but unrelated to this PR right now

Reply:
It was done per @RalfJung suggestions. Here are his comments on the matter:



    basically, if we have integer addresses, we have two cases:

        either left.layout.ty.is_integral() && right.layout.ty.is_integral(). then both sides should be ptr-to-int-cast, and we proceed with binary_int_op.
        or bin_op == mir::BinOp::Offset, then call M::ptr_op.

    if we dont have integer addresses, behavior should be unchanged.

Reply:
Well now I can't look any more because the diff is broken. :/

But I somehow didn't like going through Miri and back when we have ptr-int-casts available. The code flow would be somewhat weird: Miri would do `force_bits` and then call back to here, and endless recursion would be prevented because things would then not be pointers any more.

Notice that I specifically said "if we dont have integer addresses, behavior should be unchanged" in my plan; the idea was that we would have some way to detect that (maybe check the `force` functions return value for `Err`?). But maybe you are right and it makes more sense to do all of that in Miri.

I just really don't like seeing any value-based dispatching in the code here. Dispatching should be purely type-based...